### PR TITLE
Strip unused generic type context for VASigCookie

### LINKED ISSUE #98977: [dotnet-sdk-9.0.100-preview.3.24123.6] Windows SDK thrown "Non-blittable generic types cannot be marshaled" exception.
### Description

**Description:**
When running 3rd party application with the latest .NET 9 build, applications cannot launch, or some functions don't work anymore. After investigating, we found the issue is caused by Windows sdk, which is thrown "Non-blittable generic types cannot be marshaled" exception.

**Application Name:** NatTypeTester, MyMoneyNet, FlowLauncher, ModernFlyouts, TextGrab, lively,CountdownNumberPuzzleSolver
**OS:** Windows 10 21H2
**CPU:** X64
**.NET Build Number:** dotnet-sdk-9.0.100-preview.3.24123.6
**App & Source Location checking at:** https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1977811

**Github Link:** 
https://github.com/HMBSbige/NatTypeTester
https://github.com/MoneyTools/MyMoney.Net
https://github.com/Flow-Launcher/Flow.Launcher

### Reproduction Steps

**Minimal Repro steps (Demo attached):[MarshalDirectiveExceptionDemo.zip](https://github.com/dotnet/runtime/files/14415670/MarshalDirectiveExceptionDemo.zip)**
The machine has dotnet-sdk-9.0.100-preview.3.24123.6 installed.
1.Create a new Console project.
2.Update TargetFramework in .csproject to "net9.0-windows10.0.22621.0", which is look like this:
```<TargetFramework>net9.0-windows10.0.22621.0</TargetFramework>```
3.Add following code into Program.cs file:
```
using Windows.Globalization;
using Windows.Media.Ocr;

List<Language> possibleOCRLanguages = OcrEngine.AvailableRecognizerLanguages.ToList();

// See https://aka.ms/new-console-template for more information
Console.WriteLine("Hello, World!");
```
6. Launch the app.

Expected Result: 

### Expected behavior

App launch successfully.

### Actual behavior

**Actual Result:** 
Launch failed with error:
```
System.Runtime.InteropServices.MarshalDirectiveException
  HResult=0x80131535
  Message=Non-blittable generic types cannot be marshaled.
  Source=WinRT.Runtime
  StackTrace:
   at ABI.System.Collections.Generic.IEnumerator`1.get_HasCurrent()
   at ABI.System.Collections.Generic.IEnumerator`1.FromAbiHelper.MoveNext()
   at ABI.System.Collections.Generic.IEnumerator`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.<Main>$(String[] args) in E:\Demos\MarshalDirectiveExceptionDemo\MarshalDirectiveExceptionDemo\Program.cs:line 4
```

### Regression?

**Yes**

**Verify Scenarios:** 
1). Windows 10 21H2 AMD64 + dotnet-sdk-8.0.101: Pass
2). Windows 10 21H2 AMD64 + dotnet-sdk-9.0.100-preview.2.24119.3: Pass
3). Windows 10 21H2 AMD64 + dotnet-sdk-9.0.100-preview.3.24123.6: Fail
4). Windows 10 21H2 AMD64 + dotnet-sdk-9.0.100-preview.3.24126.3: Fail

### Known Workarounds

_No response_

### Configuration

**Dotnet Info:**
.NET SDK:
 Version:           9.0.100-preview.3.24123.6
 Commit:            884fddb17a
 Workload version:  9.0.100-manifests.06e7bcac
 MSBuild version:   17.10.0-preview-24122-01+d4cb14fe4

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-preview.3.24123.6\

.NET workloads installed:
 [wasm-tools]
   Installation Source: VS 17.8.34601.278
   Manifest Version:    9.0.0-preview.2.24123.1/9.0.100-preview.2
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.2\microsoft.net.workload.mono.toolchain.current\9.0.0-preview.2.24123.1\WorkloadManifest.json
   Install Type:              Msi

Host:
  Version:      9.0.0-preview.2.24123.1
  Architecture: x64
  Commit:       99b76018b6

.NET SDKs installed:
  9.0.100-preview.3.24123.6 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 9.0.0-preview.2.24121.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 9.0.0-preview.2.24123.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 9.0.0-preview.2.24122.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

### Other information

@dotnet-actwx-bot @dotnet/compat

### PR DESCRIPTION
**Pull request by jkotas**
Fixes #98977

### CODE REVIEW

File: src/coreclr/vm/ceeload.cpp [hunk 1]
```diff
@@ -4659,6 +4659,113 @@ PTR_VOID ReflectionModule::GetRvaField(RVA field) // virtual
...
+                IfFailThrow(sp.GetData(&argCnt)); // Get number of parameters
+                while (argCnt--)
+                {
+                    if (TypeSignatureContainsGenericVariables(sp))
+                        return true;
+                }
+            }
+            return false;
+
+        default:
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ceeload.cpp:4737]:
```suggestion
        case ELEMENT_TYPE_VAR:
        case ELEMENT_TYPE_MVAR:
        default:
```

I think explicitly including the two types we want to identify makes this clearer.

File: src/coreclr/vm/ceeload.cpp [hunk 2]
```diff
@@ -4659,6 +4659,113 @@ PTR_VOID ReflectionModule::GetRvaField(RVA field) // virtual
...
+                    }
+
+                    uint32_t nlbounds;
+                    IfFailThrow(sp.GetData(&nlbounds)); // Get # of lower bounds
+                    while (nlbounds--)
+                    {
+                        IfFailThrow(sp.GetData(NULL)); // Skip lower bounds
+                    }
+                }
+
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/ceeload.cpp:4718]:
```suggestion
```

File: src/tests/nativeaot/SmokeTests/PInvoke/PInvoke.cs [hunk 3]
```diff
@@ -346,6 +346,7 @@ public static int Main()
             TestForwardDelegateWithUnmanagedCallersOnly();
             TestDecimal();
             TestDifferentModopts();
+            TestGenericCaller<string>();
```
Comment (AaronRobinsonMSFT) [inline src/tests/nativeaot/SmokeTests/PInvoke/PInvoke.cs:349]:
```suggestion
            TestGenericCaller<int>();
            TestGenericCaller<string>();
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Comment by MichalStrehovsky** [comment]
Note that we don't run the nativeaot test tree for any other flavors than nativeaot itself. So it's not really testing a change to the VM.

**Review by AaronRobinsonMSFT — APPROVED** [review]

File: src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs [hunk 4]
```diff
@@ -40,6 +46,11 @@ internal static unsafe BlittableGeneric<T> WrappedGenericCalli<U>(void* fnptr, U
         {
             return ((delegate* unmanaged<U, BlittableGeneric<T>>)fnptr)(arg);
         }
+
+        internal static unsafe void NonGenericCalli<U>(void* fnptr, ref int val, float arg)
+        {
+            ((delegate* unmanaged<float, ref int, void>)fnptr)(ref val, arg);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs:52]:
```suggestion
            ((delegate* unmanaged<ref int, float, void>)fnptr)(ref val, arg);
```

If the arguments should be the other way, we might want a comment.

File: src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs [hunk 5]
```diff
@@ -81,6 +92,14 @@ public static void RunGenericFunctionPointerTest(float inVal)
             outVar = GenericCaller<string>.WrappedGenericCalli((delegate* unmanaged<float, BlittableGeneric<string>>)&UnmanagedExportedFunctionBlittableGenericString, inVal).X;
         }
         Assert.Equal(expectedValue, outVar);
+
+        outVar = 0;
+        Console.WriteLine("Testing non-GenericCalli with non-blittable argument in a generic caller");
+        unsafe
+        {
+            GenericCaller<string>.NonGenericCalli<string>((delegate* unmanaged<float, int*, void>)&UnmanagedExportedFunctionRefInt, ref outVar, inVal);
```
Comment (AaronRobinsonMSFT) [inline src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs:100]:
```suggestion
            GenericCaller<string>.NonGenericCalli<string>((delegate* unmanaged<int*, float, void>)&UnmanagedExportedFunctionRefInt, ref outVar, inVal);
```

File: src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs (hunk 4 above)
Comment (jkotas) [inline src/tests/Interop/MarshalAPI/FunctionPointer/GenericFunctionPointer.cs:52]:
```suggestion
            ((delegate* unmanaged<ref int, float, void>)fnptr)(ref val, arg);
```

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**